### PR TITLE
Support for LSP/tsserver Code Actions

### PR DESCRIPTION
**Pull request by daliusd**
Related issue #1466

### CODE REVIEW

File: doc/ale.txt [hunk 1]
```diff
@@ -3106,6 +3106,18 @@ ALERename                                                           *ALERename*
   in those files will not be updated.
 
 
+ALECodeAction                                                   *ALECodeAction*
+
+  Code Actions support for `tsserver`.
+
+  There are two different kind of actions supported for `tsserver`. If run in
```
Comment (w0rp) [inline doc/ale.txt:3113]: There are some typos here to fix.

File: doc/ale.txt [hunk 2]
```diff
@@ -3106,6 +3106,18 @@ ALERename                                                           *ALERename*
...
+  normal mode then code fix will be attempted if there is error in that line.
+  If there are multiple code fixes available use will be shown input to choose
+  one. In visual mode `tsserver` will be queries for applicable refactors
+  (e.g. extract to constant or extract to function) and user will be given
+  choice to select the one he/she likes.
```
Comment (w0rp) [inline doc/ale.txt:3118]: I tend to just write "they" as a gender neutral term.

File: autoload/ale/completion.vim [hunk 3]
```diff
@@ -617,6 +617,7 @@ function! ale#completion#ParseLSPCompletions(response) abort
         let l:user_data = {'_ale_completion_item': 1}
 
         if has_key(l:item, 'additionalTextEdits')
+        \ && l:item.additionalTextEdits isnot v:null
```
Comment (w0rp) [inline autoload/ale/completion.vim:620]:
```suggestion
        \&& l:item.additionalTextEdits isnot v:null
```

Preference is to not add a space.

File: autoload/ale/codefix.vim [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+    else
+        let [l:line, l:column] = getpos("'<")[1:2]
+        let [l:end_line, l:end_column] = getpos("'>")[1:2]
+    endif
+
+    let l:column = min([l:column, len(getline(l:line))])
+    let l:end_column = min([l:end_column, len(getline(l:end_line))])
+
+    let l:Callback = function(
+    \ 's:OnReady', [l:line, l:column, l:end_line, l:end_column])
```
Comment (w0rp) [inline autoload/ale/codefix.vim:366]: I tend to put the closing parenthesis on its own line.

**Review by w0rp — APPROVED** [review]
This is great. Let's just do it.

I'll modify the things I've mentioned here.

**Comment by w0rp** [comment]
Cheers! :beers:

File: autoload/ale/code_action.vim [hunk 5]
```diff
@@ -33,35 +33,35 @@ endfunction
 
 function! s:ChangeCmp(left, right) abort
     if a:left.start.line < a:right.start.line
-        return -1
+        return 1
     endif
```
[the remark this answers is not in the document]
Comment (daliusd) [reply autoload/ale/code_action.vim:37]: Sounds good.

File: doc/ale.txt (hunk 2 above)
Comment (daliusd) [reply doc/ale.txt:3118]: Yes. That sounds better.

**Comment by daliusd** [comment]
Cheers! :beers:

**Pull request merged**